examples/AffinePlaneCurve.jl

```diff
@@ -0,0 +1,148 @@
+module AffinePlaneCurveModule
+
+using Oscar, Markdown
+
+export AffinePlaneCurve, degree, jacobi_ideal, tangent, issmooth_point, union,
+       Point
+
+################################################################################
+#
+# We follow the lecture notes by Andreas Gathmann: 
+# https://www.mathematik.uni-kl.de/~gathmann/en/curves.php
+#
+################################################################################
+mutable struct AffinePlaneCurve{S}
+  eq::Oscar.MPolyElem{S}                  # Equation of the curve (polynomial in two variables)
+  degree::Int64                           # degree of the equation of the curve
+  dimension::Int64                        # dimension of the curve (as a variety)
+  function AffinePlaneCurve(eq::Oscar.MPolyElem{S}) where {S <: FieldElem}
+    r = new{S}()
+    r.eq = eq
+    if nvars(parent(r.eq)) != 2           
+       error("The defining equation must belong to a ring with two variables")
+    else
+       r.degree = -1                      # -1 when it is not computed yet
+       r.dimension = 1                    # since C is a plane curve, the dimension is always 1
+       return r
+    end
+  end
+end
+
+function Base.show(io::IO, C::AffinePlaneCurve)
+  println(io, "Affine plane curve defined by ", C.eq)
+end
+
+################################################################################
+# Point (first attempt).
+# The point is described by its coordinates.
+# ambient_dim gives the dimension of the space to which belongs the point.
+
+mutable struct Point{S}
+  coord::Array{S, 1}
+  ambient_dim::Int64
+  function Point(coordinates::Array{S,1}) where {S <: FieldElem}
+    r = new{S}()
+    r.coord = coordinates
+    r.ambient_dim = length(coordinates)
+    return r
+  end
+end
+
+function Base.show(io::IO, P::Point)
+  println(io, "Point with coordinates ", P.coord)
+end
+
+################################################################################
+# Compute the degree of the equation of the curve if not already known, and show it.
+
+@doc Markdown.doc"""
+degree(C::AffinePlaneCurve)
+> Given an affine plane curve C, returns the degree of the defining polynomial.
+"""
+function Oscar.degree(C::AffinePlaneCurve)
+  if C.degree >= 0
+     return C.degree
+  else
+     C.degree = total_degree(C.eq)
+     return(C.degree)
+  end
+end
+
+################################################################################
+# Compute the Jacobian Ideal of C
+
+@doc Markdown.doc"""
+jacobi_ideal(C::AffinePlaneCurve)
+> Given an affine plane curve C, returns the Jacobian ideal of the defining polynomial.
+"""
+function Oscar.jacobi_ideal(C::AffinePlaneCurve)
+ return jacobi_ideal(C.eq)
+end
+
+################################################################################
+# To check if a point is smooth: return true if the point is a smooth point on the curve C, and false  if it is singular, and an error if it is not on the curve.
```

Comment:
Some of these comments could be spread over two lines. They look like they are currently running over the end of the line on an 80 column terminal.